crates/matrix-sdk/src/widget/machine/mod.rs

```diff
@@ -324,17 +339,23 @@ impl WidgetMachine {
                 let (request, action) = self.send_matrix_driver_request(request);
 
                 request.then(|result, machine| {
-                    let response = result.and_then(|mut events| {
-                        let CapabilitiesState::Negotiated(capabilities) = &machine.capabilities
-                        else {
-                            let err = "Received read event request before capabilities negotiation";
-                            return Err(err.into());
-                        };
-
-                        events.retain(|e| capabilities.raw_event_matches_read_filter(e));
-                        Ok(ReadEventResponse { events })
-                    });
-
+                    let response = match (result, &machine.capabilities) {
+                        (Ok(mut events), CapabilitiesState::Negotiated(capabilities)) => {
+                            events.retain(|e| capabilities.raw_event_matches_read_filter(e));
+                            Ok(ReadEventResponse { events })
+                        }
+                        (Ok(_), CapabilitiesState::Unset) => {
+                            Err(FromWidgetErrorResponse::from_string(
+                                "Received read event request before capabilities negotiation",
+                            ))
+                        }
+                        (Ok(_), CapabilitiesState::Negotiating) => {
+                            Err(FromWidgetErrorResponse::from_string(
+                                "Received read event request while capabilities were negotiating",
+                            ))
+                        }
+                        (Err(e), _) => Err(FromWidgetErrorResponse::from_error(&e)),
+                    };
```

Comment:
Still think it would be slightly clearer to handle the capabilities error before the other.